File: golang/queens_parallel.go
// Copyright 2016 John Hurst
// John Hurst (john.b.hurst@gmail.com)
// 2017-12-18

package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

type Board struct {
	size   int
  row    int
	cols   int // 64-bits
	diags1 int
	diags2 int
}

func New(size int) Board {
	return Board{size: size}
}

func Ok(board Board, col int) bool {
  return board.cols & (1 << uint(col)) == 0 &&
    board.diags1 & (1 << uint(board.row + col)) == 0 &&
    board.diags2 & (1 << uint(board.row - col + board.size - 1)) == 0
}

func Place(board Board, col int) Board {
  return Board{
    size: board.size,
    row: board.row + 1,
    cols: board.cols | (1 << uint(col)),
    diags1: board.diags1 | (1 << uint(board.row + col)),
    diags2: board.diags2 | (1 << uint(board.row - col + board.size - 1))}
}

func Solve(board Board) int {
  if board.row == board.size {
    return 1
  } else {
    result := 0
    for col := 0; col < board.size; col++ {
      if Ok(board, col) {
        result += Solve(Place(board, col))
      }
    }
    return result
  }
}

func SolveForCol(board Board, col int, channel chan int) {
   channel <- Solve(Place(board, col))
}

func SolveSize(size int) int {
  chans := make([]chan int, size)
  board := New(size)
  for col := 0; col < size; col++ {
    chans[col] = make(chan int)
    go SolveForCol(board, col, chans[col])
  }
  result := 0
  for col := 0; col < size; col++ {
    result += <-chans[col]
  }
  return result
}

func Usage() {
	fmt.Printf("queens_all size1 [size2]\n")
	os.Exit(1)
}

func main() {
	if len(os.Args) < 2 || len(os.Args) > 3 {
		Usage()
	}
	from, err1 := strconv.Atoi(os.Args[1])
	to, err2 := strconv.Atoi(os.Args[len(os.Args)-1])
	if err1 != nil || err2 != nil || from > to || from < 4 || to < 4 {
		Usage()
	}

	for size := from; size <= to; size++ {
		start := time.Now()
		count := SolveSize(size)
		duration := time.Now().Sub(start).Seconds()
		fmt.Printf("%d,%d,%0.3f\n", size, count, duration)
	}
}

File: elixir/lib/QueensParallel.ex
# Copyright 2018 John Hurst
# JOhn Hurst (john.b.hurst@gmail.com)
# 2018-12-22

defmodule QueensParallel do
  def solve_main(board) do
    1..board.size
      |> Enum.map(fn col -> Task.async(Queens, :solve, [Queens.place(board, col)]) end)
      |> Enum.map(fn task -> Task.await(task, 10 * 60 * 60 * 1000) end)
      |> Enum.sum
  end

  def main(argv) do
    from = if length(argv) > 0, do: String.to_integer(Enum.at(argv, 0)), else: 8
    to = if length(argv) > 1, do: String.to_integer(Enum.at(argv, 1)), else: from
    for size <- from..to do
      start = Time.utc_now
      result = solve_main(%Board{size: size})
      duration = Time.diff(Time.utc_now, start, 1000)/1000
      IO.puts "#{size},#{result},#{duration}"
    end
  end
end
